### DISCUSSION TITLE
Making a mids object

### DISCUSSION THREAD

**Discussion by dr-jojo** [amices/mice#412] [Q&A] [answered]
This may be a simple question but I was unable to confirm when searching for this independently. Let's say I have multiple data frames that represent different variations of my data such as those that have been imputed. Just as a hypothetical let's say I have several excel files that are the result of imputation from a different program. Of course I can take those and make them an s3 object. But if I wanted to combine those data frames in a mids object, how would that be done? The as.mids function seems closest but not exactly what I am talking about. Any help would be appreciated.

**Comment by gerkovink** [comment, marked as answer]
This should do it
``` r
set.seed(123)
library(mice)

# mimic seperate sets
imp <- mice(nhanes, print = FALSE)
A <- complete(imp, 1)
B <- complete(imp, 2)
C <- complete(imp, 3)

# prepare long
long <- rbind(nhanes, A, B, C)
long$imp.nr <- c(rep(0:3, c(rep(25, 4))))

# invoke mids class
new.imp <- as.mids(long, .imp = "imp.nr")

# checks
is.mids(new.imp)
#> [1] TRUE
complete(new.imp, 1) ==  A
#>        age  bmi  hyp  chl
#>  [1,] TRUE TRUE TRUE TRUE
#>  [2,] TRUE TRUE TRUE TRUE
#>  [3,] TRUE TRUE TRUE TRUE
#>  [4,] TRUE TRUE TRUE TRUE
#>  [5,] TRUE TRUE TRUE TRUE
#>  [6,] TRUE TRUE TRUE TRUE
#>  [7,] TRUE TRUE TRUE TRUE
#>  [8,] TRUE TRUE TRUE TRUE
#>  [9,] TRUE TRUE TRUE TRUE
#> [10,] TRUE TRUE TRUE TRUE
#> [11,] TRUE TRUE TRUE TRUE
#> [12,] TRUE TRUE TRUE TRUE
#> [13,] TRUE TRUE TRUE TRUE
#> [14,] TRUE TRUE TRUE TRUE
#> [15,] TRUE TRUE TRUE TRUE
#> [16,] TRUE TRUE TRUE TRUE
#> [17,] TRUE TRUE TRUE TRUE
#> [18,] TRUE TRUE TRUE TRUE
#> [19,] TRUE TRUE TRUE TRUE
#> [20,] TRUE TRUE TRUE TRUE
#> [21,] TRUE TRUE TRUE TRUE
#> [22,] TRUE TRUE TRUE TRUE
#> [23,] TRUE TRUE TRUE TRUE
#> [24,] TRUE TRUE TRUE TRUE
#> [25,] TRUE TRUE TRUE TRUE
```

<sup>Created on 2021-06-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Reply by dr-jojo** [reply]
That works, thank you so much!

**Reply by dr-jojo** [reply]
I was able to use your example, but then when I try to use my own data I get an error. 

#c1, c2, and c3 are dataframes from excel files that I uploaded 

set.seed(123)
library(mice)



# prepare long
dlong = rbind(c1, c2, c3)
dlong$nr <- c(rep(1:3, c(rep(50, 3))))

#everything is good up until the next step

# invoke mids class
dlong.mid <- as.mids(dlong, .imp = "nr")

#when I actually try to use the as.mids() function this is the error I get: "Error in as.mids(dlong, .imp = "nr") : Original data not found.
 Use `complete(..., action = 'long', include = TRUE)` to save original data."

**Comment by gerkovink** [comment]
You need the original data with missings (indicated by `.imp = 0`) in order to create a valid `mids`.

**Reply by dr-jojo** [reply]
So then, you cannot make a mids object if you have only complete data?

**Reply by gerkovink** [reply]
Strictly speaking, a mids object contains the original data and a list of imputed values for each missing datum. If you only have three imputed sets, you could figure out which data is likely to be observed and which data is not. The caveat lies in the imputations that are identical over sets. They could erroneously be considered observed, which will impact the between imputation variance. To avoid this, it is better to require the user to specify the observed set. 

In short, you need to figure out which cells are observed and which are not and feed that to `as.mids()`too

**Reply by dr-jojo** [reply]
Okay, thanks. I should probably email you and explain more of what I am trying to do if that is okay.

**Reply by gerkovink** [reply]
I have a good feeling of what you'd like to do. See my comment below. Let me know when I am mistaken.

**Comment by gerkovink** [comment]
To illustrate why it may be challenging to distill an observed set from imputed sets:

``` r
set.seed(123)
library(mice)
imp <- mice(nhanes, printFlag = FALSE)

A <- complete(imp, 1)
B <- complete(imp, 2)
C <- complete(imp, 3)

R1 <- A == B
R2 <- B == C
R3 <- A == C #redundant

R <- R1 & R2 & R3
obs <- A
obs[R == FALSE] <- NA

identical(obs, nhanes)
#> [1] FALSE
obs[10:12, ]
#>    age bmi hyp chl
#> 10   2  NA   1  NA
#> 11   1  NA   1  NA
#> 12   2  NA  NA  NA
nhanes[10:12, ]
#>    age bmi hyp chl
#> 10   2  NA  NA  NA
#> 11   1  NA  NA  NA
#> 12   2  NA  NA  NA
```

<sup>Created on 2021-07-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by gerkovink** [comment]
To create a mids object from imputations only:

``` r
set.seed(123)
library(mice)
library(dplyr)
library(purrr)
imp <- mice(nhanes, 
            printFlag = FALSE, 
            where = matrix(TRUE, nrow(nhanes), ncol(nhanes)))

# make list of completed data
com <- complete(imp, "all") # this list contains all imputed sets

# make empty mids object based on first imputed set
test <- mice(com[[1]], printFlag = FALSE)

# fill imp$imp in test
c.names <- names(test$data)
for(i in 1:length(c.names)){
  test$imp[[c.names[i]]] <- com %>% map(c.names[i]) %>% do.call(cbind, .)
}
# indicate that all values are imputations
test$where <- !test$where
test$data[test$where] <- NA

# test
com.new <- complete(test, "all")
identical(com, com.new)
#> [1] TRUE

# inferences 
com %>% map(lm, formula = bmi ~ age + chl) %>% pool()
#> Class: mipo    m = 5 
#>          term m    estimate         ubar            b            t dfcom
#> 1 (Intercept) 5 20.67804557 1.651515e+01 4.710849e+01 73.045335185    22
#> 2         age 5 -0.38894869 9.347895e-01 1.579740e+00  2.830478021    22
#> 3         chl 5  0.03425162 5.278468e-04 8.000913e-04  0.001487956    22
#>         df      riv    lambda       fmi
#> 1 2.715497 3.422930 0.7739055 0.8530218
#> 2 3.820589 2.027931 0.6697415 0.7665831
#> 3 4.109094 1.818917 0.6452538 0.7450545
com.new %>% map(lm, formula = bmi ~ age + chl) %>% pool()
#> Class: mipo    m = 5 
#>          term m    estimate         ubar            b            t dfcom
#> 1 (Intercept) 5 20.67804557 1.651515e+01 4.710849e+01 73.045335185    22
#> 2         age 5 -0.38894869 9.347895e-01 1.579740e+00  2.830478021    22
#> 3         chl 5  0.03425162 5.278468e-04 8.000913e-04  0.001487956    22
#>         df      riv    lambda       fmi
#> 1 2.715497 3.422930 0.7739055 0.8530218
#> 2 3.820589 2.027931 0.6697415 0.7665831
#> 3 4.109094 1.818917 0.6452538 0.7450545
```

<sup>Created on 2021-07-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

[edited] Made `test$data` completely `NA` to avoid any potential confusion

**Reply by dr-jojo** [reply]
Thanks again!
